pkg/bufuuid/uuid.go

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bufuuid
+
+import (
+	"crypto/rand"
+	"io"
+
+	"github.com/google/uuid"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+const uuidLen = uint64(len(uuid.UUID{}))
+
+// Generator provides convenience methods to generate random v4 UUIDs, leveraging
+// an internal buffer to amortize the cost of reading from the random source.
+type Generator struct {
+	mu     lock.Mutex
+	rander io.Reader
+
+	buffer []byte
+	cursor uint64
+}
+
+// New returns a new object that can be used to generate random v4 UUID instances,
+// analogously to "github.com/google/uuid".New(). It leverages an internal buffer
+// to amortize the cost of reading from the random source when generating a large
+// number of UUIDs. This is similar to "github.com/google/uuid".EnableRandPool,
+// but confined to this specific generator instance, so that it can be used when
+// security is not a concern (i.e., it is not a problem that the buffer is stored
+// on the Go heap), without needing to enable it globally.
+func New() *Generator {
+	// The default number of slots has been selected based on BenchmarkUUIDGenerator
+	// considering that the amount of memory overhead is negligible (1024B).
+	//
+	// goos: linux
+	// goarch: amd64
+	// pkg: github.com/cilium/cilium/pkg/bufuuid
+	// cpu: 12th Gen Intel(R) Core(TM) i7-12700H
+	// BenchmarkUUIDGenerator/1_slots-20         	 4218040	       283.7 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/2_slots-20         	 8177311	       147.2 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/4_slots-20         	11446687	       101.8 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/8_slots-20         	17418846	        71.12 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/16_slots-20        	20795576	        53.49 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/32_slots-20        	26962387	        44.19 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/64_slots-20        	30800746	        40.13 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/128_slots-20       	29712133	        38.75 ns/op	       0 B/op	       0 allocs/op
+	// BenchmarkUUIDGenerator/256_slots-20       	30200934	        36.99 ns/op	       0 B/op	       0 allocs/op
+	const slots = 64
+	return NewCustom(rand.Reader, slots)
+}
+
+func NewCustom(rander io.Reader, slots uint64) *Generator {
+	return &Generator{
+		rander: rander,
+		buffer: make([]byte, slots*uuidLen),
+	}
+}
+
+// New creates a new random UUID or panics.
+func (g *Generator) New() uuid.UUID {
+	var u uuid.UUID
+	g.NewInto(&u)
+	return u
+}
+
+// NewInto writes a new random UUID into target, or panics.
+func (g *Generator) NewInto(target *uuid.UUID) {
+	g.mu.Lock()
+
+	if g.cursor == 0 || g.cursor == uint64(cap(g.buffer)) {
```

Comment:
Is this more efficient than "just" wrapping the rander in a `bufio.Reader`?

Reply:
Hmm, let me double-check. I had tested using the upstream `uuid.NewRandomFromReader` in combination with `bufio.Reader`, which has the drawback of still requiring an allocation for every call. But there it may work just fine.

Reply:
Yeah, performance wise it is practically equivalent:

```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/bufuuid
cpu: 13th Gen Intel(R) Core(TM) i7-13800H

BenchmarkUUIDGenerator/1_slots-20                3633464               323.4 ns/op             0 B/op          0 allocs/op
BenchmarkUUIDGenerator/2_slots-20                7148876               169.9 ns/op             0 B/op          0 allocs/op
BenchmarkUUIDGenerator/4_slots-20                9964256               117.9 ns/op             0 B/op          0 allocs/op
BenchmarkUUIDGenerator/8_slots-20               15018026                81.28 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDGenerator/16_slots-20              19537220                61.67 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDGenerator/32_slots-20              23108659                51.50 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDGenerator/64_slots-20              24541465                47.08 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDGenerator/128_slots-20             26875652                44.30 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDGenerator/256_slots-20             27294463                42.97 ns/op            0 B/op          0 allocs/op

BenchmarkUUIDBufioGenerator/1_slots-20           3750579               321.0 ns/op             0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/2_slots-20           7027934               169.4 ns/op             0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/4_slots-20           9728568               120.5 ns/op             0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/8_slots-20          14477647                81.09 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/16_slots-20         18793405                63.79 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/32_slots-20         20183020                54.58 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/64_slots-20         24100729                49.63 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/128_slots-20        25398668                46.94 ns/op            0 B/op          0 allocs/op
BenchmarkUUIDBufioGenerator/256_slots-20        25767868                45.73 ns/op            0 B/op          0 allocs/op
```

Where the `BufferedGenerator` implementation is:

```
func (g *BufferedGenerator) NewInto(target *uuid.UUID) {
	g.mu.Lock()
	io.ReadFull(g.rander, target[:])
	g.mu.Unlock()

	target[6] = (target[6] & 0x0f) | 0x40 // Version 4
	target[8] = (target[8] & 0x3f) | 0x80 // Variant is 10
}
```

and `g.rander = bufio.NewReaderSize(rander, int(slots*uuidLen))`

I'll switch to implementing it on top of `bufio.Reader` directly then.

Reply:
Updated, thanks for the suggestion. Note that the benchmark results are worse compared to the ones mentioned in the previous version of the commit, as performed on a different machine.